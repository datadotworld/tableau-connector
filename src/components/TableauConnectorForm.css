/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */

.main {
  margin-top: 2rem;
  padding: 0 0.5rem;
}

.header-image {
  display: block;
  margin-right: auto;
  margin-left: auto;
}

.header {
  text-align: center;
  font-size: 1.75rem;
}

.main.query form {
  margin-top: 0.5rem;
}

.main.query .header-image{
  display: inline-block;
  margin-right: 1rem;
  width: 50px;
  height: 50px;
}

.main.query .header {
  display: inline;
  vertical-align: middle;
}

.footer {
  margin-top: 1rem;
  text-align: center;
}

.buttonBar {
  text-align: center;
}

.buttonBar .btn + .btn {
  margin-left: 2rem;
}

button.button-addon-with-icon {
  padding-right: 2.5rem;
}

.browse-icon {
  width: 20px;
  fill: #6290c3;
  margin-bottom: -9px;
  margin-right: 5px;
}

.btn-default:hover .browse-icon {
  fill: #4376af;
}