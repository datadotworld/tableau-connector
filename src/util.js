import axios from 'axios'
import crypto from 'crypto'

const codeVerifierKey = 'DW-CODE-VERIFIER'

function getCodeVerifier() {
  return window.localStorage.getItem(codeVerifierKey)
}

function generateCodeChallenge(codeVerifier) {
  const base64hash = crypto.createHash('sha256')
    .update(codeVerifier)
    .digest('base64')
  return encodeURIComponent(base64hash)
}

export function storeCodeVerifier() {
  window.localStorage.setItem(codeVerifierKey, generateCodeVerifier())
}

export function removeCodeVerifier() {
  window.localStorage.removeItem(codeVerifierKey)
}

export function generateCodeVerifier() {
  const lowerCase = 'abcdefghijklmnopqrstuvwxyz'
  const upperCase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  const numbers = '1234567890'
  const specialCharacters = "-._~"
  const characters = lowerCase + upperCase + numbers + specialCharacters

  const minLength = 43
  const maxLength = 128
  const stringLength = Math.floor(Math.random() * (maxLength - minLength + 1)) + minLength

  let codeVerifier = ''
  for (var i = 0; i < stringLength; i++) {
    const randomIndex = Math.floor(Math.random() * characters.length)
    codeVerifier += characters.charAt(randomIndex)
  }

  return codeVerifier;
}

export function getAuthUrl() {
  const codeVerifier = getCodeVerifier()
  const codeChallenge = generateCodeChallenge(codeVerifier)

  return `https://data.world/oauth/authorize?client_id=${process.env.REACT_APP_OAUTH_CLIENT_ID}&redirect_uri=${process.env.REACT_APP_OAUTH_REDIRECT_URI}&response_type=code&code_challenge_method=S256&code_challenge=${codeChallenge}`
}

export function getToken(code) {
  return axios.post('https://data.world/oauth/access_token', {
    code,
    client_id: process.env.REACT_APP_OAUTH_CLIENT_ID,
    client_secret: process.env.REACT_APP_OAUTH_CLIENT_SECRET,
    grant_type: 'authorization_code',
    code_verifier: getCodeVerifier()
  });
}
