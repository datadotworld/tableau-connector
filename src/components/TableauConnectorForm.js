/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import './TableauConnectorForm.css'
import {
  Alert,
  Button,
  Col,
  ControlLabel,
  FormControl,
  FormGroup,
  Grid,
  HelpBlock,
  InputGroup,
  Row
} from 'react-bootstrap'
import sparkle from '../static/img/new-sparkle-logo.png'
import analytics from '../analytics'

const datasetRegex = /^https?:\/\/data\.world\/(.+\/.+)$/
class TableauConnectorForm extends Component {

  static propTypes = {
    connector: PropTypes.any,
    dataset: PropTypes.string,
    apiKey: PropTypes.string,
    query: PropTypes.string,
    queryType: PropTypes.string,
    clearApiKey: PropTypes.func,
    clearStoredData: PropTypes.func
  }

  state = {
    dataset: this.props.dataset || '',
    apiToken: this.props.apiKey || '',
    query: this.props.query || '',
    queryType: this.props.queryType || 'sql',
    writingQuery: !!this.props.query,
    isSubmitting: false,
    errorMessage: ''
  }

  componentDidMount = () => {
    if (this.props.apiKey) {
      analytics.identify(this.props.apiKey)
    }
  }

  datasetChanged = (e) => {
    const dataset = e.target.value.toLowerCase()
    this.setState({ dataset })
  }

  queryChanged = (e) => {
    const query = e.target.value
    this.setState({ query })
  }

  queryTypeChanged = (e) => {
    const queryType = e.target.value
    this.setState({ queryType })
  }

  isDatasetValid = () => {
    return this.state.dataset && datasetRegex.test(this.state.dataset)
  }

  onSubmit = (e) => {
    analytics.track('tableauconnector.form.submit')
    e.preventDefault()

    if (!this.isDatasetValid()) {
      return this.setState({
        isError: true
      })
    }

    if (this.state.query) {
      return this.onQuery(e)
    }

    this.setState({
      isSubmitting: true,
      isError: false
    })
    this.props.connector.setConnectionData(this.state.dataset.match(datasetRegex)[1], this.state.apiToken)
    this.props.connector.verify().then(() => {
      this.props.connector.submit()
      this.props.clearStoredData();
    }).catch((error) => {
      if (error.response && error.response.status === 401) {
        return this.props.clearApiKey()
      }
      this.setState({
        isSubmitting: false,
        isError: true,
        errorMessage: error.response && error.response.data
      })
    })
  }

  onQuery = (e) => {
    if (!this.state.query) {
      this.setState({
        isError: true
      })
    } else {
      this.setState({
        isSubmitting: true,
        isError: false
      })
      this.props.connector.setConnectionData(this.state.dataset.match(datasetRegex)[1], this.state.apiToken, this.state.query, this.state.queryType)
      this.props.connector.verify().then(() => {
        this.props.connector.submit()
        this.props.clearStoredData();
      }).catch((error) => {
        if (error.response && error.response.status === 401) {
          return this.props.clearApiKey()
        } else {
          this.setState({
            isSubmitting: false,
            isError: true,
            errorMessage: error.response && error.response.data
          })
        }
      })
    }
  }

  supportLinkClick = () => {
    analytics.track('tableauconnector.form.support.click')
  }

  render () {
    const { dataset, isSubmitting, isError, writingQuery, errorMessage } = this.state

    let datasetValidState

    if (dataset) {
      datasetValidState = this.isDatasetValid() ? 'success' : 'warning'
    }

    return (
      <Grid className={writingQuery ? 'query main' : 'main'}>
        <Row className='center-block'>
          <Col md={6} mdOffset={3} xs={10} xsOffset={1}>
            <img src={sparkle} className='header-image' alt='data.world sparkle logo' />
            <h2 className='header'>Add a data source from data.world</h2>
            <form onSubmit={this.onSubmit}>
              {isError && <Alert bsStyle='danger'>
                <strong>
                  {!errorMessage && <span>All fields are required.</span>}
                  {errorMessage && <span>{errorMessage}</span>}
                </strong>
              </Alert>}
              <FormGroup validationState={datasetValidState}>
                <ControlLabel>Dataset</ControlLabel>
                <InputGroup>
                  <FormControl
                    onChange={this.datasetChanged}
                    value={this.state.dataset}
                    autoFocus
                    type='text'
                    placeholder='http://data.world/jonloyens/an-intro-to-dataworld-dataset' />
                </InputGroup>
                {datasetValidState === 'warning' && <HelpBlock>A valid dataset URL is required: https://data.world/jonloyens/an-intro-to-dataworld-dataset</HelpBlock>}
                {datasetValidState === 'success' && <HelpBlock>Dataset URL valid</HelpBlock>}
                {!datasetValidState && <HelpBlock>Copy and paste the dataset URL here</HelpBlock>}
              </FormGroup>
              {writingQuery && <div><FormGroup>
                <ControlLabel>Query Type</ControlLabel>
                <InputGroup>
                  <FormControl
                    value={this.state.queryType}
                    onChange={this.queryTypeChanged}
                    componentClass='select'
                    placeholder='Select'>
                    <option value='sql'>SQL</option>
                    <option value='sparql'>SPARQL</option>
                  </FormControl>
                </InputGroup>
                </FormGroup>
                <FormGroup>
                  <ControlLabel>Query</ControlLabel>
                  <InputGroup>
                    <FormControl
                      onChange={this.queryChanged}
                      value={this.state.query}
                      componentClass='textarea'
                      placeholder='SELECT * FROM TABLE_NAME' />
                  </InputGroup>
                </FormGroup></div>}
              <Button
                className='center-block'
                type='submit'
                disabled={isSubmitting || datasetValidState !== 'success'}
                bsStyle='primary'>Submit</Button>
              <div className='footer'>
                <a href='https://datadotworld.zendesk.com/hc/en-us/articles/115010298907-Tableau-Web-Data-Connector' target='_blank' onClick={this.supportLinkClick}>Learn more about the data.world connector</a>
              </div>
            </form>
          </Col>
        </Row>
      </Grid>
    )
  }
}

export default TableauConnectorForm
