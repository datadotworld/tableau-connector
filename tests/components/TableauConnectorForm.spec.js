/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */

import '../util'
import renderer from 'react-test-renderer'

import React from 'react'
import ReactDOM from 'react-dom'
import TableauConnectorForm from '../../src/components/TableauConnectorForm'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<TableauConnectorForm />, div)
})

it('renders form', () => {
  expect(renderer.create(<TableauConnectorForm />).toJSON()).toMatchSnapshot()
})

it('renders form with default values', () => {
  expect(renderer.create(<TableauConnectorForm dataset='https://data.world/test/test' apiKey='testtest-testtest'/>).toJSON()).toMatchSnapshot()
})

it('renders form with query input, sql selected', () => {
  expect(renderer.create(<TableauConnectorForm dataset='https://data.world/test/test' apiKey='testtest-testtest' query='example sql query' queryType='sql'/>).toJSON()).toMatchSnapshot()
})

it('renders form with query input, sparql selected', () => {
  expect(renderer.create(<TableauConnectorForm dataset='https://data.world/test/test' apiKey='testtest-testtest' query='example sparql query' queryType='sparql'/>).toJSON()).toMatchSnapshot()
})