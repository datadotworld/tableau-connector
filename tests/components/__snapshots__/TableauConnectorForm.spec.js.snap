exports[`test renders form 1`] = `
<div
  className="main container">
  <div
    className="center-block row">
    <div
      className="col-md-6 col-md-offset-3 col-xs-10 col-xs-offset-1">
      <img
        alt="data.world sparkle logo"
        className="header-image"
        src="new-sparkle-logo.png" />
      <h2
        className="header">
        Add a data source from data.world
      </h2>
      <form
        onSubmit={[Function]}>
        <div
          className="form-group">
          <label
            className="control-label"
            htmlFor={undefined}>
            Dataset
          </label>
          <span
            className="input-group">
            <input
              autoFocus={true}
              className="form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="http://data.world/jonloyens/an-intro-to-dataworld-dataset"
              type="text"
              value="" />
          </span>
          <span
            className="help-block">
            Copy and paste the dataset URL here
          </span>
        </div>
        <button
          className="center-block btn btn-primary"
          disabled={true}
          type="submit">
          Submit
        </button>
        <div
          className="footer">
          <a
            href="https://datadotworld.zendesk.com/hc/en-us/articles/115010298907-Tableau-Web-Data-Connector"
            onClick={[Function]}
            target="_blank">
            Learn more about the data.world connector
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`test renders form with default values 1`] = `
<div
  className="main container">
  <div
    className="center-block row">
    <div
      className="col-md-6 col-md-offset-3 col-xs-10 col-xs-offset-1">
      <img
        alt="data.world sparkle logo"
        className="header-image"
        src="new-sparkle-logo.png" />
      <h2
        className="header">
        Add a data source from data.world
      </h2>
      <form
        onSubmit={[Function]}>
        <div
          className="form-group has-success">
          <label
            className="control-label"
            htmlFor={undefined}>
            Dataset
          </label>
          <span
            className="input-group">
            <input
              autoFocus={true}
              className="form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="http://data.world/jonloyens/an-intro-to-dataworld-dataset"
              type="text"
              value="https://data.world/test/test" />
          </span>
          <span
            className="help-block">
            Dataset URL valid
          </span>
        </div>
        <button
          className="center-block btn btn-primary"
          disabled={false}
          type="submit">
          Submit
        </button>
        <div
          className="footer">
          <a
            href="https://datadotworld.zendesk.com/hc/en-us/articles/115010298907-Tableau-Web-Data-Connector"
            onClick={[Function]}
            target="_blank">
            Learn more about the data.world connector
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`test renders form with query input, sparql selected 1`] = `
<div
  className="query main container">
  <div
    className="center-block row">
    <div
      className="col-md-6 col-md-offset-3 col-xs-10 col-xs-offset-1">
      <img
        alt="data.world sparkle logo"
        className="header-image"
        src="new-sparkle-logo.png" />
      <h2
        className="header">
        Add a data source from data.world
      </h2>
      <form
        onSubmit={[Function]}>
        <div
          className="form-group has-success">
          <label
            className="control-label"
            htmlFor={undefined}>
            Dataset
          </label>
          <span
            className="input-group">
            <input
              autoFocus={true}
              className="form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="http://data.world/jonloyens/an-intro-to-dataworld-dataset"
              type="text"
              value="https://data.world/test/test" />
          </span>
          <span
            className="help-block">
            Dataset URL valid
          </span>
        </div>
        <div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Query Type
            </label>
            <span
              className="input-group">
              <select
                className="form-control"
                id={undefined}
                onChange={[Function]}
                placeholder="Select"
                type={undefined}
                value="sparql">
                <option
                  value="sql">
                  SQL
                </option>
                <option
                  value="sparql">
                  SPARQL
                </option>
              </select>
            </span>
          </div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Query
            </label>
            <span
              className="input-group">
              <textarea
                className="form-control"
                id={undefined}
                onChange={[Function]}
                placeholder="SELECT * FROM TABLE_NAME"
                type={undefined}
                value="example sparql query" />
            </span>
          </div>
        </div>
        <button
          className="center-block btn btn-primary"
          disabled={false}
          type="submit">
          Submit
        </button>
        <div
          className="footer">
          <a
            href="https://datadotworld.zendesk.com/hc/en-us/articles/115010298907-Tableau-Web-Data-Connector"
            onClick={[Function]}
            target="_blank">
            Learn more about the data.world connector
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`test renders form with query input, sql selected 1`] = `
<div
  className="query main container">
  <div
    className="center-block row">
    <div
      className="col-md-6 col-md-offset-3 col-xs-10 col-xs-offset-1">
      <img
        alt="data.world sparkle logo"
        className="header-image"
        src="new-sparkle-logo.png" />
      <h2
        className="header">
        Add a data source from data.world
      </h2>
      <form
        onSubmit={[Function]}>
        <div
          className="form-group has-success">
          <label
            className="control-label"
            htmlFor={undefined}>
            Dataset
          </label>
          <span
            className="input-group">
            <input
              autoFocus={true}
              className="form-control"
              id={undefined}
              onChange={[Function]}
              placeholder="http://data.world/jonloyens/an-intro-to-dataworld-dataset"
              type="text"
              value="https://data.world/test/test" />
          </span>
          <span
            className="help-block">
            Dataset URL valid
          </span>
        </div>
        <div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Query Type
            </label>
            <span
              className="input-group">
              <select
                className="form-control"
                id={undefined}
                onChange={[Function]}
                placeholder="Select"
                type={undefined}
                value="sql">
                <option
                  value="sql">
                  SQL
                </option>
                <option
                  value="sparql">
                  SPARQL
                </option>
              </select>
            </span>
          </div>
          <div
            className="form-group">
            <label
              className="control-label"
              htmlFor={undefined}>
              Query
            </label>
            <span
              className="input-group">
              <textarea
                className="form-control"
                id={undefined}
                onChange={[Function]}
                placeholder="SELECT * FROM TABLE_NAME"
                type={undefined}
                value="example sql query" />
            </span>
          </div>
        </div>
        <button
          className="center-block btn btn-primary"
          disabled={false}
          type="submit">
          Submit
        </button>
        <div
          className="footer">
          <a
            href="https://datadotworld.zendesk.com/hc/en-us/articles/115010298907-Tableau-Web-Data-Connector"
            onClick={[Function]}
            target="_blank">
            Learn more about the data.world connector
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;
